File: scripts/loopclose.R
#!/usr/bin/env Rscript
# loopclose.R -- Register drift-offset scan fragments into one closed cloud.
# R port of loopclose.py, for a Python-free RStudio workflow.
#
# WHAT THIS IS FOR (same problem as loopclose.py)
# ------------------------------------------------
# When a single mobile-LiDAR capture (e.g. an ARKit-based scanner) loses
# SLAM tracking partway through a circumferential pass around an object, the
# wrap doesn't close -- the start/end edges (or a mid-loop band) image the
# same surface but sit offset by accumulated drift. Cut the offending arcs
# into separate fragment PLYs (e.g. in CloudCompare), then use this script
# to register and merge them back into one cloud.
#
# HOW THIS DIFFERS FROM loopclose.py -- READ BEFORE RELYING ON THIS
# -------------------------------------------------------------------
# loopclose.py needs no initial guess: it coarse-aligns each fragment with
# RANSAC global registration on FPFH features, THEN refines with ICP. R has
# no equivalent to Open3D's FPFH/RANSAC global registration -- nothing in
# this project's R dependency chain (including Rvcg) provides feature-based,
# initial-guess-free registration. This script instead skips a coarse step
# entirely and refines directly with plain point-to-point rigid ICP
# (Kabsch/SVD) on the fragments' AS-EXPORTED coordinates, correspondences
# capped at --max-corr-dist exactly like the Python version's guard against
# a partial flap dragging onto the wrong surface.
#
# This relies on fragments cut from the SAME continuous ARKit capture
# already sharing a coordinate frame -- the drift this script fixes is a
# moderate accumulated offset, not an arbitrary unknown pose, so as-exported
# coordinates are already a reasonable ICP starting point. (An earlier
# version of this script added a centroid-matching "coarse" step before
# ICP; testing showed that actively HURTS here, since two different partial
# arcs of the same ring have different centroids even with zero drift
# between them -- centroid matching assumes near-total overlap, which is
# exactly the case this script does NOT have. Removed rather than kept as a
# false safety net.)
#
# Because there's no feature-based global search, this is NOT a general-
# purpose registration method: it can only recover a good alignment when the
# fragments are already close in the same frame. If a fragment pair starts
# far apart, mis-rotated, or mirrored, this script can converge to the wrong
# alignment without an obvious error (fitness/RMS can look plausible while
# being wrong) -- sanity-check the actual merged geometry (e.g. reopen
# stitched/<tag>.ply in CloudCompare) before trusting a join this script
# made, more so than you would for the Python version. If you have Python
# available, prefer loopclose.py; use this only when you don't.
#
# Usage
# -----
#   Rscript scripts/loopclose.R arcA.ply arcB.ply --tag mytree
#   Rscript scripts/loopclose.R "work/mytree_*.ply" --tag mytree --voxel 0.01 --max-corr-dist 0.02
#
# INPUT LOCATION NOTE: don't read arc pieces from a read-only "raw export"
# source folder -- cut the arcs and save them to a working folder first.
# Output always goes to --out-dir (default stitched/), never back over an input.
# =============================================================================
suppressMessages(library(Rvcg))

# --------------------------------------------------------------------- args
args <- commandArgs(trailingOnly = TRUE)
get_flag <- function(name, default = NULL) {
  i <- which(args == name)
  if (length(i) == 0) return(default)
  args[i + 1]
}
has_flag <- function(name) name %in% args

if (length(args) == 0 || has_flag("--help") || has_flag("-h")) {
  cat("Usage: Rscript loopclose.R <fragment.ply> [<fragment2.ply> ...] --tag <id>\n",
      "  --voxel FLOAT          working voxel size in metres (default 0.01)\n",
      "  --max-corr-dist FLOAT  ICP max correspondence distance in metres (default 0.02)\n",
      "  --min-overlap FLOAT    warn if fitness falls below this (default 0.15)\n",
      "  --max-overlap FLOAT    warn if fitness exceeds this on a partial arc (default 0.90)\n",
      "  --out-dir PATH         output folder (default stitched/)\n")
  quit(status = if (length(args) == 0) 1 else 0)
}

flag_positions <- which(args %in% c("--tag", "--voxel", "--max-corr-dist",
                                     "--min-overlap", "--max-overlap", "--out-dir"))
value_positions <- flag_positions + 1
pieces_raw <- args[setdiff(seq_along(args), c(flag_positions, value_positions))]

tag           <- get_flag("--tag")
voxel         <- as.numeric(get_flag("--voxel", "0.01"))
max_corr_dist <- as.numeric(get_flag("--max-corr-dist", "0.02"))
min_overlap   <- as.numeric(get_flag("--min-overlap", "0.15"))
max_overlap   <- as.numeric(get_flag("--max-overlap", "0.90"))
out_dir       <- get_flag("--out-dir", "stitched")

if (is.null(tag)) stop("--tag is required.")

# Expand any globs the shell (or Windows) didn't.
paths <- unlist(lapply(pieces_raw, function(p) {
  hits <- Sys.glob(p)
  if (length(hits)) hits else p
}))
missing <- paths[!file.exists(paths)]
if (length(missing)) stop("Fragment file(s) not found: ", paste(missing, collapse = ", "))
if (any(vapply(paths, function(p) {
  norm <- gsub("\\\\", "/", p)
  identical(strsplit(norm, "/")[[1]][1], "raw_ply")
}, logical(1)))) {
  stop("Refusing to read from raw_ply/ (read-only field exports). ",
       "Save your cut arcs to a working folder first.")
}
if (length(paths) < 2) stop("Need at least 2 fragment PLYs to register.")

# ----------------------------------------------------------------------- IO
read_ply_xyz <- function(path) {
  mesh <- vcgImport(path, clean = FALSE, silent = TRUE)
  xyz <- t(mesh$vb[1:3, , drop = FALSE])
  if (nrow(xyz) == 0) stop(sprintf("%s: no points read.", path))
  xyz
}

write_ply_xyz <- function(xyz, path) {
  mesh <- list(vb = rbind(t(xyz), 1), it = matrix(integer(0), nrow = 3, ncol = 0))
  class(mesh) <- "mesh3d"
  vcgPlyWrite(mesh, path, binary = TRUE)
}

# Mean-per-occupied-cell voxel downsample -- same role as Open3D's
# voxel_down_sample: caps point density for ICP speed and, at the end, dedupes
# overlap between merged fragments.
voxel_downsample <- function(xyz, voxel) {
  cell <- floor(xyz / voxel)
  key <- paste(cell[, 1], cell[, 2], cell[, 3], sep = "_")
  agg <- rowsum(xyz, key) / as.vector(table(key)[match(unique(key), names(table(key)))])
  # rowsum's row order follows sort(unique(key)); recompute cleanly via split to avoid
  # relying on rowsum's internal ordering guarantees.
  idx <- split(seq_len(nrow(xyz)), key)
  t(vapply(idx, function(ix) colMeans(xyz[ix, , drop = FALSE]), numeric(3)))
}

# --------------------------------------------------------- rigid ICP (Kabsch)
# Closed-form optimal rotation+translation between two matched Nx3 point
# sets (Kabsch/Umeyama SVD solution, reflection-corrected). Point-to-point,
# not point-to-plane -- the one further simplification versus loopclose.py's
# TransformationEstimationPointToPlane, kept for simplicity since this
# script's coarse step is already the bigger fidelity gap.
kabsch <- function(src, tgt) {
  src_c <- colMeans(src); tgt_c <- colMeans(tgt)
  H <- t(src - matrix(src_c, nrow(src), 3, byrow = TRUE)) %*%
       (tgt - matrix(tgt_c, nrow(tgt), 3, byrow = TRUE))
  sv <- svd(H)
  d <- sign(det(sv$v %*% t(sv$u)))
  R <- sv$v %*% diag(c(1, 1, d)) %*% t(sv$u)
  t <- tgt_c - as.vector(R %*% src_c)
  list(R = R, t = t)
}

apply_transform <- function(xyz, tf) {
  t(tf$R %*% t(xyz) + tf$t)
}

# Iterative closest point, correspondences capped at max_corr_dist (the same
# guard loopclose.py's docstring explains -- without it, a small partial
# fragment gets dragged onto non-overlapping surface to satisfy a false
# 100%-overlap assumption). Returns the cumulative transform plus final
# fitness (inlier fraction) / inlier RMS, computed the same way as Open3D
# reports them.
icp_align <- function(src, tgt, max_corr_dist, max_iter = 60, tol = 1e-7) {
  cur <- src
  R_tot <- diag(3); t_tot <- c(0, 0, 0)
  prev_rmse <- Inf
  fitness <- 0; inlier_rmse <- NA_real_

  for (it in seq_len(max_iter)) {
    nn <- vcgKDtree(tgt, cur, k = 1)
    d <- as.vector(nn$distance)
    inliers <- d < max_corr_dist
    fitness <- mean(inliers)
    if (!any(inliers)) break
    inlier_rmse <- sqrt(mean(d[inliers]^2))

    tf <- kabsch(cur[inliers, , drop = FALSE], tgt[nn$index[inliers], , drop = FALSE])
    cur <- apply_transform(cur, tf)
    R_tot <- tf$R %*% R_tot
    t_tot <- as.vector(tf$R %*% t_tot + tf$t)

    if (abs(prev_rmse - inlier_rmse) < tol) break
    prev_rmse <- inlier_rmse
  }
  list(R = R_tot, t = t_tot, fitness = fitness, inlier_rmse = inlier_rmse)
}

# --------------------------------------------------------------------- driver
cat(sprintf("Loop-closing %d fragments for tree %s (R port -- direct ICP, no RANSAC/FPFH coarse step):\n",
            length(paths), tag))
for (p in paths) cat("  -", p, "\n")
cat(sprintf("  voxel=%.0f mm   max_corr_dist=%.0f mm\n\n", voxel * 1000, max_corr_dist * 1000))

body <- read_ply_xyz(paths[1])
body_down <- voxel_downsample(body, voxel)
logs <- list()

for (p in paths[-1]) {
  src <- read_ply_xyz(p)
  src_down <- voxel_downsample(src, voxel)

  # no coarse step -- see module header: relies on fragments already
  # sharing a coordinate frame (same ARKit capture), refines directly.
  fit <- icp_align(src_down, body_down, max_corr_dist)
  rms_mm <- 1000 * fit$inlier_rmse

  flag <- ""
  if (fit$fitness < min_overlap) {
    flag <- sprintf("  ** LOW fitness (<%.2f) -- join may not have locked **", min_overlap)
  } else if (fit$fitness > max_overlap) {
    flag <- sprintf("  ** HIGH fitness (>%.2f) on a partial arc -- check it didn't collapse onto the wrong surface (and, given this script's weaker coarse step, double-check the join visually) **", max_overlap)
  }
  cat(sprintf("[%s]  fitness=%.3f  inlier_RMS=%.1f mm%s\n",
              basename(p), fit$fitness, rms_mm, flag))
  logs[[length(logs) + 1]] <- list(fragment = basename(p),
                                    fitness = round(fit$fitness, 3),
                                    inlier_rms_mm = round(rms_mm, 2))

  # bake the ICP transform into the full-resolution fragment
  src_aligned <- apply_transform(src, list(R = fit$R, t = fit$t))
  body <- rbind(body, src_aligned)
  body_down <- voxel_downsample(body, voxel)
}

before <- nrow(body)
merged <- voxel_downsample(body, voxel)
cat(sprintf("\nMerged %s -> %s points after voxel dedupe.\n",
            format(before, big.mark = ","), format(nrow(merged), big.mark = ",")))

dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)
out_path <- file.path(out_dir, sprintf("%s.ply", tag))
write_ply_xyz(merged, out_path)
cat(sprintf("Wrote %s\n", out_path))

if (length(logs)) {
  worst <- logs[[which.max(vapply(logs, function(l) l$inlier_rms_mm, numeric(1)))]]
  cat(sprintf("\nWorst join: %s (%.1f mm, fitness %.3f)\n",
              worst$fragment, worst$inlier_rms_mm, worst$fitness))
  cat("Next: measure the band with dab_itsme.R (or fit_dab.py/fit_dab.R) and compare the\n",
      "circle-fit RMS against the manually-cleaned baseline. If registration can't tighten\n",
      "it further, this scan is a re-capture candidate -- or re-run with loopclose.py if\n",
      "Python is available, since its RANSAC/FPFH coarse step is more robust than this\n",
      "script's centroid-only alignment.\n", sep = "")
}

File: scripts/loopclose.py
#!/usr/bin/env python3
"""
loopclose.py — Register drift-offset scan fragments into one closed cloud.

When a single mobile-LiDAR capture (e.g. an ARKit-based scanner like
ForestScanner) loses SLAM tracking partway through a circumferential pass
around an object, the wrap doesn't close — the start/end edges (or a
mid-loop band) image the same surface but sit offset by accumulated drift.
The manual fix is to cut the offending arcs into separate clouds (e.g. in
CloudCompare) and align them by hand-picked point pairs. This script does
the same alignment in code, for a reproducible / batch-able cross-check.

Given N fragment PLYs (the arcs you cut out), this:
  1. voxel-downsamples + estimates normals + FPFH features on each,
  2. coarse-aligns each fragment to the growing body with **RANSAC global
     registration** on FPFH feature matches (no initial guess needed),
  3. refines with **point-to-plane ICP** using an explicit
     `--max-corr-dist` and a realistic overlap expectation,
  4. merges + voxel-dedupes into one cloud -> stitched/<tag>.ply,
  5. logs per-pair fitness + inlier RMS so a bad join is obvious.

Usage
-----
    python loopclose.py arcA.ply arcB.ply --tag mytree
    python loopclose.py work/mytree_*.ply --tag mytree --voxel 0.01 --max-corr-dist 0.02

WHY THIS EXISTS / THE GOTCHA IT ENCODES
---------------------------------------
A naive ICP registration on a small, partial fragment can make things WORSE
than not registering at all: if it assumes **100% overlap** between a small
edge flap and a near-complete body, it will drag the flap to "match"
non-overlapping surface, inflating the join error rather than closing it.
The fix is to never assume full overlap: ICP here only corresponds points
within `--max-corr-dist`, and we report `fitness` (the inlier fraction). For
a partial arc you should EXPECT a low fitness (~0.2-0.3), not 1.0 — a
suspiciously high fitness on a small flap means it collapsed onto the wrong
surface. `--min-overlap`/`--max-overlap` just warn when fitness falls
outside the band you expect; they don't force anything.

INPUT LOCATION NOTE
--------------------
Don't read arc pieces from a read-only "raw export" source folder — cut the
arcs and save them to a working folder first, then point this script at
those. Output always goes to `stitched/`, never back over an input.
"""

from __future__ import annotations
import argparse
import glob
import os
import sys

import numpy as np

try:
    import open3d as o3d
except ImportError:
    sys.exit("Missing dependency: pip install open3d  (see requirements.txt)")


# ----------------------------------------------------------------------------- IO
def load_cloud(path: str) -> o3d.geometry.PointCloud:
    pcd = o3d.io.read_point_cloud(path)
    if len(pcd.points) == 0:
        raise ValueError(f"{path}: no points read.")
    return pcd


def preprocess(pcd: o3d.geometry.PointCloud, voxel: float):
    """Downsample, estimate normals, and compute FPFH features at `voxel` scale.

    Returns (downsampled_cloud, fpfh). Normal + feature radii follow the open3d
    global-registration convention (2x and 5x the voxel size)."""
    down = pcd.voxel_down_sample(voxel)
    down.estimate_normals(
        o3d.geometry.KDTreeSearchParamHybrid(radius=voxel * 2, max_nn=30))
    fpfh = o3d.pipelines.registration.compute_fpfh_feature(
        down, o3d.geometry.KDTreeSearchParamHybrid(radius=voxel * 5, max_nn=100))
    return down, fpfh


# ------------------------------------------------------------ pairwise registration
def coarse_ransac(src_down, tgt_down, src_fpfh, tgt_fpfh, voxel: float):
    """RANSAC global registration on FPFH matches — no initial pose needed.

    `distance_threshold` at 1.5x voxel is the standard open3d recipe; correspondences
    farther than this are rejected, so a small arc can't falsely match the far wall."""
    dist = voxel * 1.5
    return o3d.pipelines.registration.registration_ransac_based_on_feature_matching(
        src_down, tgt_down, src_fpfh, tgt_fpfh, mutual_filter=True,
        max_correspondence_distance=dist,
        estimation_method=o3d.pipelines.registration.TransformationEstimationPointToPoint(False),
        ransac_n=3,
        checkers=[
            o3d.pipelines.registration.CorrespondenceCheckerBasedOnEdgeLength(0.9),
            o3d.pipelines.registration.CorrespondenceCheckerBasedOnDistance(dist),
        ],
        criteria=o3d.pipelines.registration.RANSACConvergenceCriteria(100000, 0.999))


def refine_icp(src, tgt, init_T, max_corr_dist: float):
    """Point-to-plane ICP refine, correspondences capped at `max_corr_dist`.

    The cap is the whole point (see module docstring): it is what stops a partial
    flap from being dragged onto non-overlapping bark. `tgt` must have normals."""
    if not tgt.has_normals():
        tgt.estimate_normals(
            o3d.geometry.KDTreeSearchParamHybrid(radius=max_corr_dist * 2, max_nn=30))
    return o3d.pipelines.registration.registration_icp(
        src, tgt, max_corr_dist, init_T,
        o3d.pipelines.registration.TransformationEstimationPointToPlane(),
        o3d.pipelines.registration.ICPConvergenceCriteria(max_iteration=200))


# --------------------------------------------------------------------------- driver
def loopclose(paths, tag, voxel, max_corr_dist, min_overlap, max_overlap, out_dir):
    if len(paths) < 2:
        sys.exit("Need at least 2 fragment PLYs to register.")

    print(f"Loop-closing {len(paths)} fragments for tree {tag}:")
    for p in paths:
        print(f"  - {p}")
    print(f"  voxel={voxel*1000:.0f} mm   max_corr_dist={max_corr_dist*1000:.0f} mm\n")

    # First fragment is the reference body; register each subsequent arc onto the
    # accumulated cloud, transform it into place, and merge.
    body = load_cloud(paths[0])
    body_down, body_fpfh = preprocess(body, voxel)
    merged = body           # full-resolution accumulator
    logs = []

    for i, p in enumerate(paths[1:], start=1):
        src = load_cloud(p)
        src_down, src_fpfh = preprocess(src, voxel)

        coarse = coarse_ransac(src_down, body_down, src_fpfh, body_fpfh, voxel)
        fine = refine_icp(src_down, body_down, coarse.transformation, max_corr_dist)

        rms_mm = 1000 * fine.inlier_rmse
        fit = fine.fitness
        flag = ""
        if fit < min_overlap:
            flag = f"  ** LOW fitness (<{min_overlap:.2f}) — join may not have locked **"
        elif fit > max_overlap:
            flag = (f"  ** HIGH fitness (>{max_overlap:.2f}) on a partial arc - "
                    "check it didn't collapse onto the wrong surface **")
        print(f"[{os.path.basename(p)}]  fitness={fit:.3f}  "
              f"inlier_RMS={rms_mm:.1f} mm{flag}")
        logs.append({"fragment": os.path.basename(p), "fitness": round(fit, 3),
                     "inlier_rms_mm": round(rms_mm, 2)})

        # Bake the transform into the full-res fragment and grow the body.
        src.transform(fine.transformation)
        merged += src
        body_down, body_fpfh = preprocess(merged, voxel)

    # Dedupe overlap: voxel-downsample the merged cloud at the working scale.
    before = len(merged.points)
    merged = merged.voxel_down_sample(voxel)
    print(f"\nMerged {before:,} -> {len(merged.points):,} points after voxel dedupe.")

    os.makedirs(out_dir, exist_ok=True)
    out_path = os.path.join(out_dir, f"{tag}.ply")
    o3d.io.write_point_cloud(out_path, merged)
    print(f"Wrote {out_path}")

    worst = max(logs, key=lambda r: r["inlier_rms_mm"]) if logs else None
    if worst:
        print(f"\nWorst join: {worst['fragment']} "
              f"({worst['inlier_rms_mm']:.1f} mm, fitness {worst['fitness']:.3f})")
        print("Next: measure the band with fit_dab.py and compare the circle-fit RMS "
              "against the manually-cleaned baseline. If registration can't tighten "
              "it further, this scan is a re-capture candidate.")
    return out_path, logs


def main():
    ap = argparse.ArgumentParser(
        description="Register drift-offset scan fragments into one closed cloud.")
    ap.add_argument("pieces", nargs="+",
                    help="Fragment PLYs (the arcs you cut out). Glob-expanded. "
                         "Don't pass files straight from a read-only raw-export folder.")
    ap.add_argument("--tag", required=True, help="Object/tree tag, used for the output name.")
    ap.add_argument("--voxel", type=float, default=0.01,
                    help="Working voxel size in METERS (default 0.01 = 1 cm). "
                         "Sets downsample + FPFH scale + final dedupe.")
    ap.add_argument("--max-corr-dist", type=float, default=0.02,
                    help="ICP max correspondence distance in METERS (default 0.02 = 2 cm). "
                         "Keep it tight — this is the guard against the 100%%-overlap "
                         "failure described in the module docstring.")
    ap.add_argument("--min-overlap", type=float, default=0.15,
                    help="Warn if ICP fitness (inlier fraction) falls below this.")
    ap.add_argument("--max-overlap", type=float, default=0.90,
                    help="Warn if fitness exceeds this on a partial arc (likely collapse).")
    ap.add_argument("--out-dir", default="stitched",
                    help="Output folder (default: stitched/). Never writes to raw_ply/.")
    args = ap.parse_args()

    # Expand any globs the shell didn't (Windows / quoted args).
    paths = []
    for pat in args.pieces:
        hits = sorted(glob.glob(pat))
        paths.extend(hits if hits else [pat])
    missing = [p for p in paths if not os.path.exists(p)]
    if missing:
        sys.exit("Fragment file(s) not found: " + ", ".join(missing))
    if any(os.path.normpath(p).split(os.sep)[0] == "raw_ply" for p in paths):
        sys.exit("Refusing to read from raw_ply/ (read-only field exports). "
                 "Save your cut arcs to a working folder first.")

    loopclose(paths, args.tag, args.voxel, args.max_corr_dist,
              args.min_overlap, args.max_overlap, args.out_dir)


if __name__ == "__main__":
    main()
